llvm/lib/Transforms/Scalar/GVNSink.cpp

```diff
@@ -575,6 +607,13 @@ class GVNSink {
     unsigned NumSunk = 0;
     ReversePostOrderTraversal<Function*> RPOT(&F);
     VN.setReachableBBs(BasicBlocksSet(RPOT.begin(), RPOT.end()));
+    // Populated DFSNumbers ahead of time to avoid updating dominator tree
+    // when CFG is modified. The DFSNumbers of newly created basic blocks
+    // are irrelevant because RPOT is also obtained ahead of time and only
+    // DFSNumbers of original CFG are relevant for sinkable candidates.
+    for (auto *BB : RPOT)
```

Comment:
At this point, do you actually need DFS numbering?  If any arbitrary numbering works, might as well just number in IR order.

Reply:
That's true, it'll simplify a lot of things as you noted. I'll update the patch.